net: latmon: latency monitor service and sample (#85154)
zephyrproject-rtos/zephyr: samples/net/latmon/README.rst

Reviewer: While it is not a rule I can refer to, a lot of the rst files keep a maximum of 100 line length, which I do prefer.

.. zephyr:code-sample:: latmon-client
   :name: Latmon Client
   :relevant-api: latmon

   Measures delta time between GPIO events and reports the latency metrics via latmon to the latmus
   service executing on the SUT.

Overview
********

This project provides tools to measure the worst-case response time of a system under test (SUT) to
GPIO events using:

- **Latmon**:

  Runs on a Zephyr-based board to generate and monitor GPIO events while collecting metrics

- **Latmus**:

  Runs on the SUT to respond to the falling edge of input GPIO event, displaying the latency metrics
  and generate histogram data.

This project is part of the open-source initiative
`EVL Project - Latmus GPIO Response Time <https://evlproject.org/core/benchmarks/#latmus-gpio-response-time>`_.

The main program is designed to monitor latency using GPIO pins on a Zephyr-based system. It generates a
pulse signal on a GPIO pin and measures the time it takes for the SUT (executing Latmus) to respond to
it.

The SUT must be running Latmus to capture the latency metrics and histogram information reported over the

> network. The program uses LEDS to indicate the different states, such as DHCP binding(red), waiting for the

Reviewer 2: ```suggestion
network. The program uses LEDs to indicate the different states, such as DHCP binding(red), waiting for the
```

Latmus connection (blue) and sampling (green).

Why Not Just Use a Timer?
=========================

Timer tests miss external factors like GPIO signals, hardware, and interrupt handling.
Latmon and Latmus simulate real-world scenarios, capturing end-to-end latency metrics.
This ensures accurate assessment of real-time responsiveness across the entire system.

- **Real-Time Thread Testing**:

  Evaluates how a user-space thread processes external interrupts.

- **End-to-End Latency Measurement**:

  Captures delays from hardware, drivers, and user-space threads.

- **Versatile Platform Support**:

  Works on EVL, PREEMPT_RT, and other platforms.

Code Structure
==============

The Latmon sample application is divided into two main components:

> - **Application Logic** (``samples/net/latmon/main.c``):

Reviewer 2: ```suggestion
- **Application Logic** (:zephyr_file:`samples/net/latmon/main.c`):
```

   This file contains the application logic for Latmon.
   It initializes networking, provides the instrumentation mechanism and handles LED indicators for the
   different states.

> - **Library** (``subsys/net/lib/latmon/latmon.c``):

Reviewer 2: ```suggestion
- **Library** (:zephyr_file:`subsys/net/lib/latmon/latmon.c`):
```

   This file provides reusable functions and abstractions for latency monitoring via Latmus.
   It includes the core logic for reporting latency metrics and histogram data.

Requirements
************

- **Zephyr-Compatible Board**:

  A board with external GPIO support and an IPv4 network interface (e.g., FRDM-K64F).

- **System under Test**:

  A system with external GPIO pins running the Latmus service and an IPv4 network interface.

- **Network Connection**:

  A DHCP server for IP assignment.

- **Physical Connection**:

  GPIO wires connecting the Zephyr board to the SUT and both systems connected to the network.

Setup and Usage
***************

- **Flash Latmon onto the Zephyr board**:

  The application will connect to the network and wait for a connection from the SUT. The application
  will use DHCP to obtain an IPv4 address.

- **Connect GPIO pins for transmit (Zephyr to SUT) and receive (SUT to Zephyr)**

>   On **FRMD-K64F**, the sample code uses the **Arduino header J2**, ``pin 20`` for transmit the pulse to

Reviewer 2: ```suggestion
  On **FRDM-K64F**, the sample code uses the **Arduino header J2**, ``pin 20`` for transmit the pulse to
```

  the SUT and ``pin 18`` to receive the acknowledgment from the SUT.

- **Run Latmus on the SUT**

  Request the appropriate options with `Latmus <https://evlproject.org/core/testing/#latmus-program>`_. Users
  can for example modify the sampling period with the ``-p`` option or generate historgram data for
  postprocessing with the ``-g`` option,

- **Monitor results from the SUT**

  Latmus will report latency figures and, if requested, generate the histogram data file.

- **Calibrating the Latmus latencies: CONFIG_LATMON_LOOPBACK_CALIBRATION**:

  Users can connect the GPIO pins in loopback mode (transmit to ack) and build the Latmon sample application with
  CONFIG_LATMON_LOOPBACK_CALIBRATION enabled. When connecting to Latmus in this configuration, Latmus is providing
  a calibration value that can be used to adjust the final latencies.

Example
=======

> On the host and to build and flash the Zephyr FRMD-K6F board with the Latmon sample:

Reviewer 2: ```suggestion
On the host and to build and flash the Zephyr FRDM-K64F board with the Latmon sample:
```

.. code-block:: console

>    host$ west build -b frdm_k64f samples/net/latmon

Reviewer 2: more a nit than anything but if you would use a shell prompt like e.g. `user@host:~$ ` then it will be possible for the user to copy paste commands from the docs using the "copy" button with the prompts automatically being discarded. Apply everywhere you have console blocks
